File: c++/longest_substring_without_repeating_characters.cpp
/*
Given a string, find the length of the longest substring without repeating characters.

Examples:

Given "abcabcbb", the answer is "abc", which the length is 3.

Given "bbbbb", the answer is "b", with the length of 1.

Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
*/


/*
 *
*/

#include "helper.h"

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> map;
        int maxLen = 0, start = -1;
        for (int i = 0; i < s.length(); i++) {
            if (map.find(s[i]) != map.end() && map[s[i]] > start) {
                start = map[s[i]];
            }
            map[s[i]] = i;
            maxLen = max(maxLen, i - start);
        }
        return maxLen;
    }
};

int main() {
    Solution s;
    cout << s.lengthOfLongestSubstring("abba") << endl;
    cout << s.lengthOfLongestSubstring("abcabcbb") << endl;
    cout << s.lengthOfLongestSubstring("dvdf") << endl;
    cout << s.lengthOfLongestSubstring("bbbbb") << endl;
    cout << s.lengthOfLongestSubstring("pwwkew") << endl;

    return 0;
}

File: python/longest_substring_without_repeating_characters.py
'''
Given a string, find the length of the longest substring without repeating characters. For example, the longest substring without repeating letters for "abcabcbb" is "abc", which the length is 3. For "bbbbb" the longest substring is "b", with the length of 1.
'''

'''
Use two pointers, i(current_pos) and last_pos. and use a dict visited to save the most recent position for a char. line "last_pos = max(visited[s[i]] + 1, last_pos)" is the key. When we find an existing char, only update last_pos if the most recent appearance of the char is larger than last_pos. And don't forget to calculate when iter is over and specical cases len(s) <= 1
'''
class Solution:
    # @return an integer
    def lengthOfLongestSubstring(self, s):
        if len(s) <= 1:
            return len(s)
        max_len = 0
        visited = {}
        last_pos = 0
        for i in xrange(len(s)):
            if s[i] not in visited:
                visited[s[i]] = i
            else:
                if i - last_pos > max_len:
                    max_len = i - last_pos
                last_pos = max(visited[s[i]] + 1, last_pos)
                visited[s[i]] = i
        if i + 1 - last_pos > max_len:
            max_len = i + 1 - last_pos
        return max_len

if __name__ == '__main__':
    s = Solution()
    print s.lengthOfLongestSubstring('abcdbbadc')
    print s.lengthOfLongestSubstring('bbbbb')
    print s.lengthOfLongestSubstring('wlrbbmqbhcdarzowkkyhiddqscdxrjmowfrxsjybldbefsarcbynecdyggxxpklorellnmpapqfwkhopkmco')
    print s.lengthOfLongestSubstring('qopubjguxhxdipfzwswybgfylqvjzhar')
    print s.lengthOfLongestSubstring('')
               
